packages/application/src/shell.ts

```diff
@@ -657,6 +664,25 @@ class ApplicationShell extends Widget {
     }
     this._currentChanged.emit(args);
     this._onLayoutModified();
+
+    this._onCurrentSession(args.newValue ? (args.newValue as any).session : undefined);
```

Comment:
Do we have a more type-safe way of getting at whether there is an `IClientSession` associated with a widget? This works for notebooks and consoles, but more-or-less by accident, in that they both have the `session` available by that name on the top-level widget. If another widget comes along with a different API, but still uses an `IClientSession` this will not work.

I can think of how we would do it for documents: we could use the `DocumentManager.contextForWidget` function to get the context, which has a `session` object on it. I don't currently know how to do it for a more general widget...

Reply:
Ah, sorry, I missed that you addressed this in the above comment. Yeah, I am not sure about this just now (though I think the docmanager solution would work for documents).

Reply:
Is there some way for a widget to register a `isBusy` for itself? So not base it on the session object at ll but make it more generic? Like there is the `setDirty` method on the application that widgets can use to mark if they are unsaved. The simplest way I can think of is storing a [`WeakMap`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap) in the application mapping widgets to `isBusy` signals. Any widget could add itself and a signal to this map when it is created.

Reply:
I like this idea! I think it could follow a very similar logic to the `setDirty` setup (which doesn't use a `WeakMap`, but instead a `DisposableDelegate` with an internal count of the dirty widgets.